### DISCUSSION TITLE
Index fields configuration duplication

### DISCUSSION THREAD

**Discussion by kbond** [schranz-search/schranz-search#199] [Ideas]
This is a very tiny papercut I had - not a big deal.

Consider this:

```php
use Schranz\Search\SEAL\Schema\Field;
use Schranz\Search\SEAL\Schema\Index;

return new Index('page', [
    'id' => new Field\IdentifierField('id'),
    'title' => new Field\TextField('title'),
    'subtitle1' => new Field\TextField('subtitle1'),
    'subtitle2' => new Field\TextField('subtitle2'),
    'description' => new Field\TextField('description'),
]);
```

I think it would be nice if we didn't have to set the array key:

```php
use Schranz\Search\SEAL\Schema\Field;
use Schranz\Search\SEAL\Schema\Index;

return new Index('page', [
    new Field\IdentifierField('id'),
    new Field\TextField('title'),
    new Field\TextField('subtitle1'),
    new Field\TextField('subtitle2'),
    new Field\TextField('description'),
]);
```

I could be missing something but if interested, feel free to turn into an issue and I can create a PR.

**Comment by alexander-schranz** [comment]
This is currently also related if we use the same classes maybe in a future [Object Data Mapper package](https://github.com/schranz-search/schranz-search/issues/81) or not. As it could also be:

```php
use Schranz\Search\SEAL\Schema\Field;
use Schranz\Search\SEAL\Schema\Index;

return new Index('page', [
    'id' => new Field\IdentifierField(),
    'title' => new Field\TextField(),
    'subtitle1' => new Field\TextField(),
    'subtitle2' => new Field\TextField(),
    'description' => new Field\TextField(),
]);
```

The key is currently required inside `Index.php` class as the items are currently accessed that way.

In a ODM the above example could maybe look like this:

```php
#[Index('page')]
class Test {
    #[Field\IdentifierField()]
    public string $id;
}
```

What I'm not yet sure if I should support something like doctrine does `fieldName` vs `columnName` which would be something like this:

```php
return new Index('page', [
    'fieldName' => new Field\TextField('columnName'),
]);
```

At current state that is not supported and currently avoding that to avoid to have complexity in the mapping of the stored data vs presentated data, but the initial interface was created by keep that in mind and that is way it currently looks like this.

I think it should not hurt the performance too much if we currently add support for `int` as field key but make sure that `$index->fields` always use the name as that is required to access the field. So we could adopt the `__construct` of `Index` do something like this:

```php
$normalizedFields = [];
foreach ($fields as $key => $field) {
    if (is_int($key)) {
         $key = $field->name;
    }
    
    $this->normalizedFields[$key] = $field;
}

$this->fields = $normalizedFields; // set to the readonly publi
```

This needs also keep in mind for `TypedField` and `ObjectField` which also have `fields` configuration. If you want to give it a try feel free to create a pull request for it.
